File: app/src/main/java/de/westnordost/streetcomplete/quests/housenumber/AddHousenumberForm.java
package de.westnordost.streetcomplete.quests.housenumber;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AlertDialog;
import android.text.InputType;
import android.text.TextWatcher;
import android.text.method.DigitsKeyListener;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;

import de.westnordost.streetcomplete.R;
import de.westnordost.streetcomplete.quests.AbstractQuestFormAnswerFragment;
import de.westnordost.streetcomplete.quests.building_type.BuildingType;
import de.westnordost.streetcomplete.util.TextChangedWatcher;
import de.westnordost.streetcomplete.view.Item;
import de.westnordost.streetcomplete.view.ItemViewHolder;


public class AddHousenumberForm extends AbstractQuestFormAnswerFragment
{
	public static final String
			NO_ADDRESS = "noaddress",
			HOUSENUMBER = "housenumber",
			HOUSENAME = "housename",
			CONSCRIPTIONNUMBER = "conscriptionnumber",
			STREETNUMBER = "streetnumber";

	private static final String	IS_HOUSENAME = "is_housename";
	// i.e. 9999/a, 9/a, 99/9, 99a, 99 a, 9 / a
	public static final String VALID_HOUSENUMBER_REGEX =
			"\\p{N}{1,4}((\\s?/\\s?\\p{N})|(\\s?/?\\s?\\p{L}))?";

	public static final String VALID_CONSCRIPTIONNUMBER_REGEX =	"\\p{N}{1,6}";

	@Nullable private EditText inputHouseNumber, inputHouseName, inputConscriptionNumber, inputStreetNumber;

	private boolean isHousename;

	@Override public View onCreateView(LayoutInflater inflater, ViewGroup container,
									   Bundle savedInstanceState)
	{
		View view = super.onCreateView(inflater, container, savedInstanceState);
		restoreInstanceState(savedInstanceState);

		if(isHousename)
		{
			setLayout(R.layout.quest_housename);
		}
		else
		{
			setLayout(R.layout.quest_housenumber);
		}

		addOtherAnswers();

		return view;
	}

	@Override public void onSaveInstanceState(@NonNull Bundle outState)
	{
		super.onSaveInstanceState(outState);
		outState.putBoolean(IS_HOUSENAME, isHousename);
	}

	private void restoreInstanceState(Bundle inState)
	{
		isHousename = false;
		if(inState != null)
		{
			isHousename = inState.getBoolean(IS_HOUSENAME);
		}
	}

	private void addOtherAnswers()
	{
		addOtherAnswer(R.string.quest_address_answer_no_housenumber, this::onNoHouseNumber);

		addOtherAnswer(R.string.quest_address_answer_house_name, () ->
		{
			isHousename = true;
			setLayout(R.layout.quest_housename);
		});

		addOtherAnswer(R.string.quest_housenumber_multiple_numbers, () ->
		{
			new AlertDialog.Builder(getActivity())
				.setMessage(R.string.quest_housenumber_multiple_numbers_description)
				.setPositiveButton(android.R.string.ok, null)
				.show();
		});
	}

	@Override protected void onClickOk()
	{
		if(inputHouseName != null)
		{
			applyHouseNameAnswer(getInputText(inputHouseName));
		}
		else if(inputConscriptionNumber != null && inputStreetNumber != null)
		{
			applyConscriptionNumberAnswer(getInputText(inputConscriptionNumber), getInputText(inputStreetNumber));
		}
		else if(inputHouseNumber != null)
		{
			applyHouseNumberAnswer(getInputText(inputHouseNumber));
		}
	}

	private void onNoHouseNumber()
	{
		String buildingValue = getOsmElement().getTags().get("building");
		Item item = BuildingType.getByTag("building", buildingValue);
		if(item != null)
		{
			View inner = LayoutInflater.from(getActivity()).inflate(
				R.layout.dialog_quest_address_no_housenumber, null, false);
			new ItemViewHolder(inner.findViewById(R.id.item_view)).bind(item);

			new AlertDialog.Builder(getActivity())
				.setView(inner)
				.setPositiveButton(R.string.quest_generic_hasFeature_yes, (dialog, which) -> applyNoHouseNumberAnswer())
				.setNegativeButton(R.string.quest_generic_hasFeature_no_leave_note, (dialog, which) -> onClickCantSay())
				.show();
		} else {
			// fallback in case the type of building is known by Housenumber quest but not by
			// building type quest
			onClickCantSay();
		}
	}

	private void applyNoHouseNumberAnswer()
	{
		Bundle answer = new Bundle();
		answer.putBoolean(NO_ADDRESS, true);
		applyAnswer(answer);
	}

	private void applyHouseNameAnswer(final String houseName)
	{
		Bundle answer = new Bundle();
		answer.putString(HOUSENAME, houseName);
		applyAnswer(answer);
	}

	private void applyHouseNumberAnswer(final String houseNumber)
	{
		final Bundle answer = new Bundle();
		boolean looksInvalid = !houseNumber.matches(getValidHousenumberRegex());

		confirmHousenumber(looksInvalid, () ->
		{
			answer.putString(HOUSENUMBER, houseNumber);
			applyAnswer(answer);
		});
	}

	private void applyConscriptionNumberAnswer(final String conscriptionNumber, final String streetNumber)
	{
		final Bundle answer = new Bundle();
		boolean looksInvalid = !conscriptionNumber.matches(VALID_CONSCRIPTIONNUMBER_REGEX);
		if(!streetNumber.isEmpty())
		{
			looksInvalid |= !streetNumber.matches(getValidHousenumberRegex());
		}

		confirmHousenumber(looksInvalid, () ->
		{
			answer.putString(CONSCRIPTIONNUMBER, conscriptionNumber);
			// streetNumber is optional
			if(!streetNumber.isEmpty())
			{
				answer.putString(STREETNUMBER, streetNumber);
			}
			applyAnswer(answer);
		});
	}

	@Override public boolean isFormComplete()
	{
		// streetNumber is always optional
		EditText input = getFirstNonNull(inputHouseNumber, inputHouseName, inputConscriptionNumber);
		return input != null && !getInputText(input).isEmpty();
	}

	private void setLayout(int layoutResourceId)
	{
		View view = setContentView(layoutResourceId);

		inputHouseNumber = view.findViewById(R.id.inputHouseNumber);
		inputHouseName = view.findViewById(R.id.inputHouseName);
		inputConscriptionNumber = view.findViewById(R.id.inputConscriptionNumber);
		inputStreetNumber = view.findViewById(R.id.inputStreetNumber);

		TextWatcher onChanged = new TextChangedWatcher(this::checkIsFormComplete);
		if(inputHouseNumber != null) inputHouseNumber.addTextChangedListener(onChanged);
		if(inputHouseName != null) inputHouseName.addTextChangedListener(onChanged);
		if(inputConscriptionNumber != null) inputConscriptionNumber.addTextChangedListener(onChanged);
		if(inputStreetNumber != null) inputStreetNumber.addTextChangedListener(onChanged);

		// streetNumber is always optional
		EditText input = getFirstNonNull(inputHouseNumber, inputHouseName, inputConscriptionNumber);
		if(input != null) input.requestFocus();

		initKeyboardButton(view);
	}

	private static EditText getFirstNonNull(EditText ...view)
	{
		for (EditText editText : view)
		{
			if(editText != null) return editText;
		}
		return null;
	}

	private void initKeyboardButton(View view)
	{
		final Button toggleKeyboardButton = view.findViewById(R.id.toggleKeyboard);
		if(toggleKeyboardButton != null)
		{
			toggleKeyboardButton.setText("abc");
			toggleKeyboardButton.setOnClickListener(v ->
			{
				View focus = getActivity().getCurrentFocus();
				if(focus != null && focus instanceof EditText)
				{
					EditText input = (EditText) focus;
					int start = input.getSelectionStart();
					int end = input.getSelectionEnd();
					if ((input.getInputType() & InputType.TYPE_CLASS_NUMBER) != 0)
					{
						input.setInputType(InputType.TYPE_TEXT_FLAG_NO_SUGGESTIONS);
						toggleKeyboardButton.setText("123");
					} else
					{
						input.setInputType(InputType.TYPE_CLASS_NUMBER);
						input.setKeyListener(DigitsKeyListener.getInstance("0123456789.,- /"));
						toggleKeyboardButton.setText("abc");
					}
					// for some reason, the cursor position gets lost first time the input type is set (#1093)
					input.setSelection(start, end);

					InputMethodManager imm = (InputMethodManager) getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
					imm.showSoftInput(input, InputMethodManager.SHOW_IMPLICIT);
				}
			});
		}
	}

	private String getValidHousenumberRegex()
	{
		String regex = VALID_HOUSENUMBER_REGEX;
		String additionalRegex = getCountryInfo().getAdditionalValidHousenumberRegex();
		if(additionalRegex != null)
		{
			regex = "((" + regex + ")|(" + additionalRegex + "))";
		}
		// i.e. "95-98" or "5,5a,6" etc. (but not: "1, 3" or "3 - 5" or "5,6-7")
		return "^" + regex + "((-" + regex + ")|(," + regex + ")+)?";
	}

	private static String getInputText(EditText editText)
	{
		return editText.getText().toString().trim();
	}

	private void confirmHousenumber(boolean isUnusual, final Runnable onConfirmed)
	{
		if(isUnusual)
		{
			new AlertDialog.Builder(getActivity())
					.setTitle(R.string.quest_generic_confirmation_title)
					.setMessage(R.string.quest_address_unusualHousenumber_confirmation_description)
					.setPositiveButton(R.string.quest_generic_confirmation_yes, (dialog, which) -> onConfirmed.run())
					.setNegativeButton(R.string.quest_generic_confirmation_no, null)
					.show();
		}
		else
		{
			onConfirmed.run();
		}
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/quests/housenumber/AddHousenumberForm.kt
package de.westnordost.streetcomplete.quests.housenumber

import android.content.res.ColorStateList
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import android.text.InputType
import android.text.method.DigitsKeyListener
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import android.widget.Button
import android.widget.EditText
import androidx.core.content.getSystemService
import androidx.core.view.isInvisible

import de.westnordost.streetcomplete.R
import de.westnordost.streetcomplete.quests.AbstractQuestFormAnswerFragment
import de.westnordost.streetcomplete.quests.OtherAnswer
import de.westnordost.streetcomplete.quests.building_type.BuildingType
import de.westnordost.streetcomplete.util.TextChangedWatcher
import de.westnordost.streetcomplete.view.image_select.ItemViewHolder


class AddHousenumberForm : AbstractQuestFormAnswerFragment<HousenumberAnswer>() {

    override val otherAnswers = listOf(
        OtherAnswer(R.string.quest_address_answer_no_housenumber) { onNoHouseNumber() },
        OtherAnswer(R.string.quest_address_answer_house_name) { switchToHouseName() },
        OtherAnswer(R.string.quest_housenumber_multiple_numbers) { showMultipleNumbersHint() }
    )

    private var houseNumberInput: EditText? = null
    private var houseNameInput: EditText? = null
    private var conscriptionNumberInput: EditText? = null
    private var streetNumberInput: EditText? = null
    private var blockNumberInput: EditText? = null

    private var toggleKeyboardButton: Button? = null

    private var addButton: View? = null
    private var subtractButton: View? = null

    private var isHousename: Boolean = false

    private var houseNumberInputTextColors: ColorStateList? = null

    private val isShowingHouseNumberHint: Boolean get() = houseNumberInputTextColors != null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = super.onCreateView(inflater, container, savedInstanceState)

        isHousename = savedInstanceState?.getBoolean(IS_HOUSENAME) ?: false
        setLayout(if(isHousename) R.layout.quest_housename else R.layout.quest_housenumber)

        return view
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putBoolean(IS_HOUSENAME, isHousename)
    }

    override fun onClickOk() {
        createAnswer()?.let { answer ->
            confirmHousenumber(answer.looksInvalid(countryInfo.additionalValidHousenumberRegex)) {
                applyAnswer(answer)
                if (answer.isRealHouseNumberAnswer) lastRealHousenumberAnswer = answer
            }
        }
    }

    override fun isFormComplete() = (!isShowingHouseNumberHint || isHousename) && createAnswer() != null

    /* ------------------------------------- Other answers -------------------------------------- */

    private fun switchToHouseName() {
        isHousename = true
        setLayout(R.layout.quest_housename)
    }

    private fun showMultipleNumbersHint() {
        activity?.let { AlertDialog.Builder(it)
            .setMessage(R.string.quest_housenumber_multiple_numbers_description)
            .setPositiveButton(android.R.string.ok, null)
            .show()
        }
    }

    private fun onNoHouseNumber() {
        val buildingValue = osmElement!!.tags["building"]!!
        val buildingType = BuildingType.getByTag("building", buildingValue)
        if (buildingType != null) {
            val inflater = LayoutInflater.from(requireContext())
            val inner = inflater.inflate(R.layout.dialog_quest_address_no_housenumber, null, false)
            ItemViewHolder(inner.findViewById(R.id.item_view)).bind(buildingType.item)

            AlertDialog.Builder(requireContext())
                .setView(inner)
                .setPositiveButton(R.string.quest_generic_hasFeature_yes) { _, _ -> applyAnswer(NoHouseNumber) }
                .setNegativeButton(R.string.quest_generic_hasFeature_no_leave_note) { _, _ -> composeNote() }
                .show()
        } else {
            // fallback in case the type of building is known by Housenumber quest but not by
            // building type quest
            onClickCantSay()
        }
    }

    /* -------------------------- Set (different) housenumber layout  --------------------------- */

    private fun setLayout(layoutResourceId: Int) {
        val view = setContentView(layoutResourceId)

        toggleKeyboardButton = view.findViewById(R.id.toggleKeyboardButton)
        houseNumberInput = view.findViewById(R.id.houseNumberInput)
        houseNameInput = view.findViewById(R.id.houseNameInput)
        conscriptionNumberInput = view.findViewById(R.id.conscriptionNumberInput)
        streetNumberInput = view.findViewById(R.id.streetNumberInput)
        blockNumberInput = view.findViewById(R.id.blockNumberInput)
        addButton = view.findViewById(R.id.addButton)
        subtractButton = view.findViewById(R.id.subtractButton)

        addButton?.setOnClickListener { addToHouseNumberInput(+1) }
        subtractButton?.setOnClickListener { addToHouseNumberInput(-1) }

        // must be called before registering the text changed watchers because it changes the text
        prefillBlockNumber()

        initKeyboardButton()
        // must be after initKeyboardButton because it re-sets the onFocusListener
        showHouseNumberHint()

        val onChanged = TextChangedWatcher { checkIsFormComplete() }
        houseNumberInput?.addTextChangedListener(onChanged)
        houseNameInput?.addTextChangedListener(onChanged)
        conscriptionNumberInput?.addTextChangedListener(onChanged)
        streetNumberInput?.addTextChangedListener(onChanged)
        blockNumberInput?.addTextChangedListener(onChanged)
    }

    private fun prefillBlockNumber() {
        /* the block number likely does not change from one input to the other, so let's prefill it
           with the last selected value */
        val input = blockNumberInput ?: return
        val blockNumberAnswer = lastRealHousenumberAnswer as? HouseAndBlockNumber ?: return
        input.setText(blockNumberAnswer.blockNumber)
    }

    private fun showHouseNumberHint() {
        val input = houseNumberInput ?: return
        val prev = lastRealHousenumberAnswer?.realHouseNumber ?: return

        /* The Auto fit layout does not work with hints, so we workaround this by setting the "real"
        *  text instead and make it look like it is a hint. This little hack is much less effort
        *  than to fork and fix the external dependency. We need to revert back the color both on
        *  focus and on text changed (tapping on +/- button) */
        houseNumberInputTextColors = input.textColors
        input.setTextColor(input.hintTextColors)
        input.setText(prev)
        input.addTextChangedListener(TextChangedWatcher {
            val colors = houseNumberInputTextColors
            if (colors != null) input.setTextColor(colors)
            houseNumberInputTextColors = null
        })
        input.onFocusChangeListener = View.OnFocusChangeListener { _, hasFocus ->
            updateKeyboardButtonVisibility()
            if (hasFocus) showKeyboard(input)
            val colors = houseNumberInputTextColors
            if (hasFocus && colors != null) {
                input.text = null
                input.setTextColor(colors)
                houseNumberInputTextColors = null
            }
        }
    }

    private fun addToHouseNumberInput(add: Int) {
        val input = houseNumberInput ?: return
        val prev = if (input.text.isEmpty()) {
            lastRealHousenumberAnswer?.realHouseNumber
        } else {
            input.text.toString()
        } ?: return
        val newHouseNumber = prev.addToHouseNumber(add) ?: return
        input.setText(newHouseNumber)
        input.setSelection(newHouseNumber.length)
    }

    private fun initKeyboardButton() {
        toggleKeyboardButton?.text = "abc"
        toggleKeyboardButton?.setOnClickListener {
            val focus = requireActivity().currentFocus
            if (focus != null && focus is EditText) {
                val start = focus.selectionStart
                val end = focus.selectionEnd
                if (focus.inputType and InputType.TYPE_CLASS_NUMBER != 0) {
                    focus.inputType = InputType.TYPE_TEXT_FLAG_NO_SUGGESTIONS
                    toggleKeyboardButton?.text = "123"
                } else {
                    focus.inputType = InputType.TYPE_CLASS_NUMBER
                    focus.keyListener = DigitsKeyListener.getInstance("0123456789.,- /")
                    toggleKeyboardButton?.text = "abc"
                }
                // for some reason, the cursor position gets lost first time the input type is set (#1093)
                focus.setSelection(start, end)
                showKeyboard(focus)
            }
        }
        updateKeyboardButtonVisibility()

        val onFocusChange = View.OnFocusChangeListener { v, hasFocus ->
            updateKeyboardButtonVisibility()
            if (hasFocus) showKeyboard(v)
        }
        houseNumberInput?.onFocusChangeListener = onFocusChange
        streetNumberInput?.onFocusChangeListener = onFocusChange
        blockNumberInput?.onFocusChangeListener = onFocusChange
    }

    private fun showKeyboard(focus: View) {
        val imm = activity?.getSystemService<InputMethodManager>()
        imm?.showSoftInput(focus, InputMethodManager.SHOW_IMPLICIT)
    }

    private fun updateKeyboardButtonVisibility() {
        toggleKeyboardButton?.isInvisible = !(
            houseNumberInput?.hasFocus() == true ||
            streetNumberInput?.hasFocus() == true ||
            blockNumberInput?.hasFocus() == true
        )
    }

    private fun confirmHousenumber(isUnusual: Boolean, onConfirmed: () -> Unit) {
        if (isUnusual) {
            AlertDialog.Builder(requireContext())
                .setTitle(R.string.quest_generic_confirmation_title)
                .setMessage(R.string.quest_address_unusualHousenumber_confirmation_description)
                .setPositiveButton(R.string.quest_generic_confirmation_yes) { _, _ -> onConfirmed() }
                .setNegativeButton(R.string.quest_generic_confirmation_no, null)
                .show()
        } else {
            onConfirmed()
        }
    }

    private fun createAnswer(): HousenumberAnswer? =
        if (houseNameInput != null) {
            houseNameInput?.nonEmptyInput?.let { HouseName(it) }
        }
        else if (conscriptionNumberInput != null && streetNumberInput != null) {
            conscriptionNumberInput?.nonEmptyInput?.let { conscriptionNumber ->
                val streetNumber = streetNumberInput?.nonEmptyInput // streetNumber is optional
                ConscriptionNumber(conscriptionNumber, streetNumber)
            }
        }
        else if (blockNumberInput != null && houseNumberInput != null) {
            blockNumberInput?.nonEmptyInput?.let { blockNumber ->
                houseNumberInput?.nonEmptyInput?.let { houseNumber ->
                    HouseAndBlockNumber(houseNumber, blockNumber)
                }
            }
        }
        else if (houseNumberInput != null) {
            houseNumberInput?.nonEmptyInput?.let { HouseNumber(it) }
        }
        else null

    private val EditText.nonEmptyInput:String? get() {
        val input = text.toString().trim()
        return if(input.isNotEmpty()) input else null
    }

    companion object {
        private var lastRealHousenumberAnswer: HousenumberAnswer? = null

        private const val IS_HOUSENAME = "is_housename"
    }
}

private val HousenumberAnswer.isRealHouseNumberAnswer: Boolean get() = when(this) {
    is HouseNumber -> true
    is HouseAndBlockNumber -> true
    else -> false
}

private val HousenumberAnswer.realHouseNumber: String? get() = when(this) {
    is HouseNumber -> number
    is HouseAndBlockNumber -> houseNumber
    else -> null
}

private fun String.addToHouseNumber(add: Int): String? {
    val parsed = parseHouseNumber(this) ?: return null
    when {
        add == 0 -> return this
        add > 0  -> {
            val last = when (val it = parsed.list.last()) {
                is HouseNumbersPartsRange -> it.end
                is SingleHouseNumbersPart -> it.single
            }
            return (last.number + add).toString()
        }
        add < 0  -> {
            val first = when (val it = parsed.list.first()) {
                is HouseNumbersPartsRange -> it.start
                is SingleHouseNumbersPart -> it.single
            }
            val result = first.number + add
            return if (result < 1) null else result.toString()
        }
        else -> return null
    }
}
